proposals/0000-deprecated-entities.rst

```diff
@@ -0,0 +1,112 @@
+
+Motivation
+------------
+It is a very common idiom to have a type and an identically named data constructor.
+Sometimes one would want to deprecate the use of constructor
+(for example, when using smart constructors) with a help of ``DEPRECATED`` pragma.
+However, according to the user guide, `there is currently no way to deprecate one thing without the other.
+<https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#warning-deprecated-pragma>`_
+The usual workaround would be to have a module that imports one but not the other,
+however while that's possible for the type it's not possible for the constructor.
+
+Proposed Change Specification
+-----------------------------
+
+* extend ``DEPRECATED`` pragma with a disambiguating specifier: ``data`` for data constructors and ``type`` for types.
+* the unqualified case would mean deprecating both entities as it does now.
+
+
+Effect and Interactions
+-----------------------
+An example of a use case for this is the following. Given the following module: ::
+
+    module A where
+
+    data Foo = Foo
+    {-# DEPRECATED Foo "Don't use type and data constructor Foo" #-}
+
+    data Bar = Bar
+    {-# DEPRECATED type Bar "Don't use type Bar" #-}
+
+    data Baz = Baz
+    {-# DEPRECATED data Baz "Don't use data constructor Baz" #-}
+
+When compiling the code which happens to use data constructor or type ``Foo``, we will see the following warnings: ::
+
+    Main.hs:5:8: warning: [-Wdeprecations (in -Wdefault)]
+        In the use of type constructor or class ‘Foo’ (imported from A):
+        Deprecated: "Don't use type and data constructor Foo"
+
+    Main.hs:6:7: warning: [-Wdeprecations (in -Wdefault)]
+        In the use of data constructor ‘Foo’ (imported from A):
+        Deprecated: "Don't use type and data constructor Foo"
+
+In the same vein, if we use data constructor or type ``Bar``,
+we will be warned **only when we use the type**, but not the constructor,
+since we specified which entity we want to deprecate in the above mentioned module ``A``: ::
+
+    Main.hs:8:8: warning: [-Wdeprecations (in -Wdefault)]
+        In the use of type constructor or class ‘Bar’ (imported from A):
+        Deprecated: "Don't use type Bar"
+
+Same logic applies to ``Baz``, we will be warned **only when its data constructor** is used(but not its type): ::
+
+    Main.hs:12:7: warning: [-Wdeprecations (in -Wdefault)]
+        In the use of data constructor ‘Baz’ (imported from A):
+        Deprecated: "Don't use data constructor Baz"
+
+
+When specifying several entities in one pragma,
+the sort of deprecated entity we've specified will apply to all listed entities.
+This will work - warn when these types are used(but not their constructors): ::
+
+    {-# DEPRECATED type Qux, Quux "Don't use this" #-}
+
+This will not work (parse error): ::
+
+    {-# DEPRECATED type Qux, constructor Quux "Don't use this" #-}
```

Comment:
I think it would be nicer to have the `type`/`data` qualifier apply only to the immediately-following name, such that the second example would be accepted. But I also haven't had occasion to put more than one name in a `DEPRECATE` pragma yet, so I might be wrong about the common case.